File: project-skills/service-skills-set/install-service-skills.py
#!/usr/bin/env python3
"""
Install the Service Skill Trinity into the current project.

Run from inside your target project directory:
    python3 /path/to/jaggers-agent-tools/project-skills/install-service-skills.py

Installs:
  .claude/skills/creating-service-skills/   — scaffold new service skills
  .claude/skills/using-service-skills/      — session-start catalog injection
  .claude/skills/updating-service-skills/   — drift detection on file writes
  .claude/skills/scoping-service-skills/    — task intake and service routing
  .claude/settings.json                     — SessionStart + PostToolUse hooks
  .githooks/pre-commit                      — doc-reminder (non-blocking)
  .githooks/pre-push                        — skill-staleness (non-blocking)
  .git/hooks/pre-commit + pre-push          — activated

Idempotent. Safe to re-run.
"""

import json
import shutil
import subprocess
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent.resolve()            # project-skills/service-skills-set/
REPO_ROOT  = SCRIPT_DIR.parent.parent.resolve()         # jaggers-agent-tools/
SKILLS_SRC = SCRIPT_DIR / ".claude"                     # service-skills-set/.claude/<skill>/
GIT_HOOKS  = SKILLS_SRC / "git-hooks"                   # service-skills-set/.claude/git-hooks/

GREEN  = "\033[0;32m"
YELLOW = "\033[1;33m"
NC     = "\033[0m"

TRINITY = ["creating-service-skills", "using-service-skills", "updating-service-skills", "scoping-service-skills"]

SETTINGS_HOOKS = {
    "SessionStart": [
        {"hooks": [{"type": "command",
            "command": "python3 \"$CLAUDE_PROJECT_DIR/.claude/skills/using-service-skills/scripts/cataloger.py\""}]}
    ],
    "PreToolUse": [
        {"matcher": "Read|Write|Edit|Glob|Grep|Bash",
         "hooks": [{"type": "command",
             "command": "python3 \"$CLAUDE_PROJECT_DIR/.claude/skills/using-service-skills/scripts/skill_activator.py\""}]}
    ],
    "PostToolUse": [
        {"matcher": "Write|Edit",
         "hooks": [{"type": "command",
             "command": "python3 \"$CLAUDE_PROJECT_DIR/.claude/skills/updating-service-skills/scripts/drift_detector.py\" check-hook",
             "timeout": 10}]}
    ]
}

MARKER_DOC       = "# [jaggers] doc-reminder"
MARKER_STALENESS = "# [jaggers] skill-staleness"


def get_project_root() -> Path:
    try:
        r = subprocess.run(["git", "rev-parse", "--show-toplevel"],
                           capture_output=True, text=True, check=True, timeout=5)
        root = Path(r.stdout.strip())
        if root == REPO_ROOT:
            print("Error: run this from inside your TARGET project, not jaggers-agent-tools itself.")
            sys.exit(1)
        return root
    except subprocess.CalledProcessError:
        print("Error: not inside a git repository.")
        sys.exit(1)


def install_skills(project_root: Path) -> None:
    print("\n── Skills ──────────────────────────────")
    for skill in TRINITY:
        src  = SKILLS_SRC / skill
        dest = project_root / ".claude" / "skills" / skill
        if dest.exists():
            shutil.rmtree(dest)
        shutil.copytree(src, dest, ignore=shutil.ignore_patterns("*.Zone.Identifier"))
        print(f"{GREEN}  ✓{NC} .claude/skills/{skill}/")


def install_settings(project_root: Path) -> None:
    print("\n── settings.json ───────────────────────")
    path = project_root / ".claude" / "settings.json"
    path.parent.mkdir(parents=True, exist_ok=True)

    existing = {}
    if path.exists():
        try:
            existing = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            pass

    hooks = existing.setdefault("hooks", {})
    for event, config in SETTINGS_HOOKS.items():
        if event not in hooks:
            hooks[event] = config
            print(f"{GREEN}  ✓{NC} added hook: {event}")
        else:
            print(f"{YELLOW}  ○{NC} hook already present: {event} (not overwritten)")

    path.write_text(json.dumps(existing, indent=2) + "\n", encoding="utf-8")


def install_git_hooks(project_root: Path) -> None:
    print("\n── Git hooks ───────────────────────────")
    doc_script      = GIT_HOOKS / "doc_reminder.py"
    staleness_script = GIT_HOOKS / "skill_staleness.py"

    pre_commit = project_root / ".githooks" / "pre-commit"
    pre_push   = project_root / ".githooks" / "pre-push"

    for hp in (pre_commit, pre_push):
        if not hp.exists():
            hp.parent.mkdir(parents=True, exist_ok=True)
            hp.write_text("#!/usr/bin/env bash\n", encoding="utf-8")
            hp.chmod(0o755)

    snippets = [
        (pre_commit, MARKER_DOC,
         f"\n{MARKER_DOC}\nif command -v python3 &>/dev/null && [ -f \"{doc_script}\" ]; then\n    python3 \"{doc_script}\" || true\nfi\n"),
        (pre_push, MARKER_STALENESS,
         f"\n{MARKER_STALENESS}\nif command -v python3 &>/dev/null && [ -f \"{staleness_script}\" ]; then\n    python3 \"{staleness_script}\" || true\nfi\n"),
    ]

    changed = False
    for hook_path, marker, snippet in snippets:
        content = hook_path.read_text(encoding="utf-8")
        if marker not in content:
            hook_path.write_text(content + snippet, encoding="utf-8")
            print(f"{GREEN}  ✓{NC} {hook_path.relative_to(project_root)}")
            changed = True
        else:
            print(f"{YELLOW}  ○{NC} already installed: {hook_path.name}")

    if changed:
        git_dir = project_root / ".git" / "hooks"
        git_dir.mkdir(parents=True, exist_ok=True)
        for src, name in ((pre_commit, "pre-commit"), (pre_push, "pre-push")):
            if src.exists():
                dest = git_dir / name
                shutil.copy2(src, dest)
                dest.chmod(0o755)
        print(f"{GREEN}  ✓{NC} activated in .git/hooks/")


def main() -> None:
    project_root = get_project_root()
    print(f"Installing into: {project_root}")

    install_skills(project_root)
    install_settings(project_root)
    install_git_hooks(project_root)

    print(f"\n{GREEN}Done.{NC}")
    print(f"  Hooks active: SessionStart (catalog) · PreToolUse (skill activator) · PostToolUse (drift) · pre-commit · pre-push")


if __name__ == "__main__":
    main()

File: cli/src/commands/install-service-skills.ts
import { Command } from 'commander';
import kleur from 'kleur';
import path from 'path';
import fs from 'fs-extra';
import { spawnSync } from 'child_process';

// CJS: __dirname = cli/dist/ — two levels up = package root
declare const __dirname: string;
const PKG_ROOT = path.resolve(__dirname, '../..');
const SKILLS_SRC = path.join(PKG_ROOT, 'project-skills', 'service-skills-set', '.claude');

const TRINITY = [
    'creating-service-skills',
    'using-service-skills',
    'updating-service-skills',
    'scoping-service-skills',
];

const SETTINGS_HOOKS: Record<string, unknown[]> = {
    SessionStart: [
        {
            hooks: [{
                type: 'command',
                command: 'python3 "$CLAUDE_PROJECT_DIR/.claude/skills/using-service-skills/scripts/cataloger.py"',
            }],
        },
    ],
    PreToolUse: [
        {
            matcher: 'Read|Write|Edit|Glob|Grep|Bash',
            hooks: [{
                type: 'command',
                command: 'python3 "$CLAUDE_PROJECT_DIR/.claude/skills/using-service-skills/scripts/skill_activator.py"',
            }],
        },
    ],
    PostToolUse: [
        {
            matcher: 'Write|Edit',
            hooks: [{
                type: 'command',
                command: 'python3 "$CLAUDE_PROJECT_DIR/.claude/skills/updating-service-skills/scripts/drift_detector.py" check-hook',
                timeout: 10,
            }],
        },
    ],
};

const MARKER_DOC = '# [jaggers] doc-reminder';
const MARKER_STALENESS = '# [jaggers] skill-staleness';

// ─── Pure functions (exported for testing) ───────────────────────────────────

export function mergeSettingsHooks(existing: Record<string, unknown>): {
    result: Record<string, unknown>;
    added: string[];
    skipped: string[];
} {
    const result = { ...existing };
    const hooks = (result.hooks ?? {}) as Record<string, unknown>;
    result.hooks = hooks;

    const added: string[] = [];
    const skipped: string[] = [];

    for (const [event, config] of Object.entries(SETTINGS_HOOKS)) {
        if (event in hooks) {
            skipped.push(event);
        } else {
            hooks[event] = config;
            added.push(event);
        }
    }

    return { result, added, skipped };
}

export async function installSkills(projectRoot: string, skillsSrc: string = SKILLS_SRC): Promise<{ skill: string; status: 'installed' | 'updated' }[]> {
    const results: { skill: string; status: 'installed' | 'updated' }[] = [];
    for (const skill of TRINITY) {
        const src = path.join(skillsSrc, skill);
        const dest = path.join(projectRoot, '.claude', 'skills', skill);
        const existed = await fs.pathExists(dest);
        if (existed) {
            await fs.remove(dest);
        }
        await fs.copy(src, dest, {
            filter: (src: string) => !src.includes('.Zone.Identifier'),
        });
        results.push({ skill, status: existed ? 'updated' : 'installed' });
    }
    return results;
}

export async function installGitHooks(projectRoot: string, skillsSrc: string = SKILLS_SRC): Promise<{
    hookFiles: { name: string; status: 'added' | 'already-present' }[];
}> {
    // Copy git-hook scripts into target project (no back-reference to jaggers package path)
    const gitHooksSrc = path.join(skillsSrc, 'git-hooks');
    const gitHooksDest = path.join(projectRoot, '.claude', 'git-hooks');
    await fs.copy(gitHooksSrc, gitHooksDest, { overwrite: true });

    const docScript = path.join(projectRoot, '.claude', 'git-hooks', 'doc_reminder.py');
    const stalenessScript = path.join(projectRoot, '.claude', 'git-hooks', 'skill_staleness.py');

    const preCommit = path.join(projectRoot, '.githooks', 'pre-commit');
    const prePush = path.join(projectRoot, '.githooks', 'pre-push');

    for (const hookPath of [preCommit, prePush]) {
        if (!await fs.pathExists(hookPath)) {
            await fs.mkdirp(path.dirname(hookPath));
            await fs.writeFile(hookPath, '#!/usr/bin/env bash\n', { mode: 0o755 });
        }
    }

    const snippets: [string, string, string][] = [
        [
            preCommit,
            MARKER_DOC,
            `\n${MARKER_DOC}\nif command -v python3 &>/dev/null && [ -f "${docScript}" ]; then\n    python3 "${docScript}" || true\nfi\n`,
        ],
        [
            prePush,
            MARKER_STALENESS,
            `\n${MARKER_STALENESS}\nif command -v python3 &>/dev/null && [ -f "${stalenessScript}" ]; then\n    python3 "${stalenessScript}" || true\nfi\n`,
        ],
    ];

    const hookFiles: { name: string; status: 'added' | 'already-present' }[] = [];
    let anyAdded = false;

    for (const [hookPath, marker, snippet] of snippets) {
        const content = await fs.readFile(hookPath, 'utf8');
        const name = path.basename(hookPath);
        if (!content.includes(marker)) {
            await fs.writeFile(hookPath, content + snippet);
            hookFiles.push({ name, status: 'added' });
            anyAdded = true;
        } else {
            hookFiles.push({ name, status: 'already-present' });
        }
    }

    if (anyAdded) {
        const gitHooksDir = path.join(projectRoot, '.git', 'hooks');
        await fs.mkdirp(gitHooksDir);
        for (const [src, name] of [[preCommit, 'pre-commit'], [prePush, 'pre-push']] as const) {
            if (await fs.pathExists(src)) {
                const dest = path.join(gitHooksDir, name);
                await fs.copy(src, dest, { overwrite: true });
                await fs.chmod(dest, 0o755);
            }
        }
    }

    return { hookFiles };
}

export async function installSettings(projectRoot: string): Promise<{ added: string[]; skipped: string[] }> {
    const settingsPath = path.join(projectRoot, '.claude', 'settings.json');
    await fs.mkdirp(path.dirname(settingsPath));

    let existing: Record<string, unknown> = {};
    if (await fs.pathExists(settingsPath)) {
        try {
            existing = JSON.parse(await fs.readFile(settingsPath, 'utf8'));
        } catch {
            // malformed JSON — start fresh
        }
    }

    const { result, added, skipped } = mergeSettingsHooks(existing);
    await fs.writeFile(settingsPath, JSON.stringify(result, null, 2) + '\n');
    return { added, skipped };
}

export function getProjectRoot(pkgRoot: string): string {
    const result = spawnSync('git', ['rev-parse', '--show-toplevel'], {
        encoding: 'utf8',
        timeout: 5000,
    });
    if (result.status !== 0) {
        throw new Error('Not inside a git repository. Run this command from your target project directory.');
    }
    const root = path.resolve(result.stdout.trim());
    if (root === path.resolve(pkgRoot)) {
        throw new Error('Run this from inside your TARGET project, not the jaggers-agent-tools repo itself.');
    }
    return root;
}

export function createInstallServiceSkillsCommand(): Command {
    return new Command('install-project-skill')
        .description('Install the Service Skill Trinity into the current project')
        .action(async () => {
            let projectRoot: string;
            try {
                projectRoot = getProjectRoot(PKG_ROOT);
            } catch (err) {
                console.error(kleur.red(`\n✗ ${(err as Error).message}\n`));
                process.exit(1);
            }

            console.log(kleur.dim(`\n  Installing into: ${projectRoot}\n`));

            console.log(kleur.bold('── Skills ──────────────────────────────'));
            const skillResults = await installSkills(projectRoot);
            for (const { skill, status } of skillResults) {
                const icon = status === 'installed' ? kleur.green('  ✓') : kleur.yellow('  ↺');
                console.log(`${icon} .claude/skills/${skill}/`);
            }

            console.log(kleur.bold('\n── settings.json ───────────────────────'));
            const { added, skipped } = await installSettings(projectRoot);
            for (const event of added) {
                console.log(`${kleur.green('  ✓')} added hook: ${event}`);
            }
            for (const event of skipped) {
                console.log(`${kleur.yellow('  ○')} already present: ${event} (not overwritten)`);
            }

            console.log(kleur.bold('\n── Git hooks ───────────────────────────'));
            const { hookFiles } = await installGitHooks(projectRoot);
            for (const { name, status } of hookFiles) {
                if (status === 'added') {
                    console.log(`${kleur.green('  ✓')} .githooks/${name}`);
                } else {
                    console.log(`${kleur.yellow('  ○')} already installed: ${name}`);
                }
            }
            if (hookFiles.some(h => h.status === 'added')) {
                console.log(`${kleur.green('  ✓')} activated in .git/hooks/`);
            }
            console.log(`${kleur.green('  ✓')} scripts → .claude/git-hooks/`);

            console.log(kleur.green('\n  Done.'));
            console.log(kleur.dim('  Hooks active: SessionStart · PreToolUse · PostToolUse · pre-commit · pre-push\n'));
        });
}
